hack/update-knative-components.go

```diff
@@ -0,0 +1,258 @@
+		Base:                github.Ptr("main"),
+		Head:                github.Ptr(branchName),
+		Body:                github.Ptr(body),
+		MaintainerCanModify: github.Ptr(true),
+	}
+	pr, _, err := client.PullRequests.Create(ctx, "knative", "func", &newPR)
+	if err != nil {
+		fmt.Printf("PR looks like this:\n%#v\n", pr)
+		fmt.Printf("err: %s\n", err)
+		return err
+	}
+	fmt.Println("ready")
+	return nil
+}
+
+// returns true when PR with given title already exists in knative/func repo
+// otherwise false. Returns an error if occurred, otherwise nil.
+func prExists(ctx context.Context, c *github.Client, title string) (bool, error) {
+	opt := &github.PullRequestListOptions{State: "open"}
+	list, _, err := c.PullRequests.List(ctx, "knative", "func", opt)
+	if err != nil {
+		return false, fmt.Errorf("errror pulling PRs in knative/func: %s", err)
+	}
+	for _, pr := range list {
+		if pr.GetTitle() == title {
+			// gauron99 - currently cannot update already existing PR, shouldnt happen
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// ----------------------------------------------------------------------------
+// ----------------------------------- MAIN -----------------------------------
+// ----------------------------------------------------------------------------
+
+// entry function -- essentially "func main() for this file"
+func updateComponentVersions() error {
+	prTitle := "chore: Update components' versions to latest"
+	ctx := context.Background()
+	client := github.NewClient(nil).WithAuthToken(os.Getenv("GITHUB_TOKEN"))
+
+	e, err := prExists(ctx, client, prTitle)
+	if err != nil {
+		return err
+	}
+	if e {
+		fmt.Printf("PR already exists, nothing to do, exiting")
+		return nil
+	}
+
+	projects := []struct {
+		owner, repo string
+	}{
+		{
+			owner: "knative",
+			repo:  "serving",
+		},
+		{
+			owner: "knative",
+			repo:  "eventing",
+		},
+		{
+			owner: "knative-extensions",
+			repo:  "net-contour",
+		},
+	}
+
+	// Get current versions used.
+	// Get all together to limit opening/closingthe file.
+	// Could be reworked to keep the file open through out the for cycle.
+	oldSrv, oldEvt, oldCntr, err := getVersionsFromFile()
+	if err != nil {
+		return err
+	}
+
+	updated := false
+	// cycle through all versions of components listed above, fetch their
+	// latest from github releases - cmp them - create PR for update if necessary
+	for _, p := range projects {
+		newV, err := getLatestVersion(ctx, client, p.owner, p.repo)
+		if err != nil {
+			return fmt.Errorf("error while getting latest v of %s/%s: %v", p.owner, p.repo, err)
+		}
+
+		// sync the old repo & version
+		oldV := ""
+		prefix := ""
+		switch p.repo {
+		case "serving":
+			oldV = oldSrv
+			prefix = knSrvPrefix
+		case "eventing":
+			oldV = oldEvt
+			prefix = knEvtPrefix
+		case "net-contour":
+			oldV = oldCntr
+			prefix = knCtrPrefix
+		}
+		// try and overwrite the file with new versions
+		isNew, err := tryUpdateFile(prefix, newV, oldV)
```

Comment:
Here, the versions struct would be updated with the latest versions, and then serialized out (overwriting versions.json), followed by regenerating the shell script.